### ISSUE TITLE
variable scoping in grouping when `by` char vec

### ISSUE THREAD

**Issue by jangorecki** [Rdatatable/data.table#3178]
when providing `by` as a character vector IMO we should treat those character values as columns of a data.table, not variables in upper scope
```r
library(data.table)
v5="z"
data.table(v1=1)[, .N, by="v5"]
#   v5 N
#1:  z 1
```
Of course such answer should be valid for `data.table(v1=1)[, .N, by=.(v5)]` because here we expect `v5` to be scoped from parent frame.

**Comment by MichaelChirico** [comment]
@jangorecki I am tempted to close due to breaking change, WDYT?

**Comment by jangorecki** [comment]
I think it is worth to have this breaking change to maintain quality of downstream packages.
If suddenly new column appears in the data, and  ew column name is same as variable, then workflow will silently process differently. Scope from where the value comes will be changed.

This is a usage pattern that should be avoided. I would add a warning for now.

Usage of character like this should be wrapped into get(). This is explicitly telling that variable is not in the DT scope.

**Comment by Mukulyadav2004** [comment]
Hi, although I understand this is a significant breaking change, but I'm willing to help implement the warning approach you've outlined if that would be useful.

**Comment by jangorecki** [comment]
Yes, warning would be good, but best to keep it for 1.19.0 so will be more time for downstream packages to notice new warning, if they rely on this scoping behavior.

**Comment by jangorecki** [comment]
> Of course such answer should be valid (as it is now) for `data.table(v1=1)[, .N, by=.(v5)]` because here we expect `v5` to be scoped from parent frame

After second thought I don't think it should be valid. Rather by=c(v5) should be valid only. Otherwise .(v5) resolves to list("z") and this when typed by hand gives (or at least should give) completely different results.

So here would be nice to have a warning as well.

**Comment by Mukulyadav2004** [comment]
Got it! Both cases - by as character vector as well symbol in list constructor currently look up variables in the parent scope when they should treat them as column references:
